routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,459 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package actions
+
+import (
+	"bytes"
+	"compress/gzip"
+	gocontext "context"
+	"crypto/md5"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/storage"
+	"code.gitea.io/gitea/modules/web"
+)
+
+const artifactXTfsFileLengthHeader = "x-tfs-filelength"
+const artifactXActionsResultsMD5Header = "x-actions-results-md5"
+
+// const artifactXActionsResultsCRC64Header = "x-actions-results-crc64"
+
+const artifactRouteBase = "/_apis/pipelines/workflows/{taskID}/artifacts"
+
+func ArtifactsRoutes(goctx gocontext.Context, prefix string) *web.Route {
+	m := web.NewRoute()
+	m.Use(withContexter(goctx))
+
+	r := artifactRoutes{
+		prefix: prefix,
+		fs:     storage.ActionsArtifacts,
+	}
+
+	// retrieve, list and confirm artifacts
+	m.Post(artifactRouteBase, r.getUploadArtifactURL)
+	m.Get(artifactRouteBase, r.listArtifacts)
+	m.Patch(artifactRouteBase, r.comfirmUploadArtifact)
+
+	// handle container artifacts
+	m.Put(artifactRouteBase+"/{artifactID}/upload", r.uploadArtifact)
+	m.Get(artifactRouteBase+"/{artifactID}/path", r.getDownloadArtifactURL)
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) buildArtifactUrl(taskID int64, artifactID int64, suffix string) (string, error) {
```

Comment:
Maybe we also need to convert `Url` to `URL` since other functions did that.

Reply:
updated